src/main/io/rcdevice.c

```diff
@@ -0,0 +1,1013 @@
+
+                if (isFoundANullTeminatedChar) {
+                    expecteRemainingDataLen--;
+                    if (expecteRemainingDataLen == 0) {
+                        break;;
+                    }
+                }
+
+                if (settingType == RCDEVICE_PROTOCOL_SETTINGTYPE_TEXT_SELECTION) {
+                    if (!receiveCurrentValueForTextSelectionType) {
+                        receiveCurrentValueForTextSelectionType = true;
+                    } else {
+                        if (c == 0) {
+                            // found the null teminated char, and get next char(crc field), then break the loop
+                            isFoundANullTeminatedChar = true; 
+                            expecteRemainingDataLen = 1;
+                        }
+                    }
+                } else if (settingType == RCDEVICE_PROTOCOL_SETTINGTYPE_STRING) {
+                    // found the null teminated char, and get max string size and crc field, then break the loop
+                    isFoundANullTeminatedChar = true; 
+                    expecteRemainingDataLen = 2;
+                }
+            }
+
+            if (dataPos >= expectedPacketLen) { 
+                break;
+            }
+        }
+    }
+
+    if (crc != 0) {
+        return false;
+    }
+
+    if (outputBufferLen && (*outputBufferLen >= (dataPos - 1)) && outputBuffer) {
+        memcpy(outputBuffer, data,
+               dataPos - 1); // return the data expect the crc field
+        *outputBufferLen = dataPos - 1;
+    }
+
+    return true;
+}
+
+// every time send packet to device, and want to get something from device,
+// it'd better call the method to clear the rx buffer before the packet send,
+// else may be the useless data in rx buffer will cause the response decoding
+// failed.
+static void runcamDeviceFlushRxBuffer(runcamDevice_t *device)
+{
+    while (serialRxBytesWaiting(device->serialPort) > 0) {
+        serialRead(device->serialPort);
+    }
+}
+
+// a common way to send packet to device
+static void runcamDeviceSendPacket(runcamDevice_t *device, uint8_t command, uint8_t *paramData, int paramDataLen)
+{
+    // is this device open?
+    if (!device->serialPort) {
+        return;
+    }
+
+    sbuf_t buf;
+    // prepare pointer
+    buf.ptr = device->buffer;
+    buf.end = ARRAYEND(device->buffer);
+
+    sbufWriteU8(&buf, RCDEVICE_PROTOCOL_HEADER);
+    sbufWriteU8(&buf, command);
+
+    if (paramData) {
+        sbufWriteData(&buf, paramData, paramDataLen);
+    }
+
+    // add crc over (all) data
+    crc8_dvb_s2_sbuf_append(&buf, device->buffer);
+
+    // switch to reader
+    sbufSwitchToReader(&buf, device->buffer);
+
+    // send data if possible
+    serialPort_t *serialPort = device->serialPort;
+    if (!serialPort->locked) {
+        serialPort->locked = true;
+        serialWriteBuf(device->serialPort, sbufPtr(&buf), sbufBytesRemaining(&buf));
+        serialPort->locked = false;
```

Comment:
What is this locking of the serial port trying to do? It seems highly suspect to me.

Reply:
yes, this code is from Simon(fish pepper),  we are using openTCO in fews days ago, so these locking logic is from fishpepper/betaflight/osd_device branch